README.md
# Little Lisp Interpreter in Rust

At this point this is just a poorly written calculator.


The `main.py` is what I was originally trying (huge failure). Turns out it's
much easier to do in Rust with all the types.

## TODO
- [ ] REFACTOR
  - [ ] Think about why all the `clones` and think of something better.
  - [x] Get rid of all the unnecessary `match`es.
  - [x] Get rid of all the unnecessary copies.
  - [ ] Be careful where to borrow and where to take ownership.
  - [x] Make everything functional.
  - [ ] Find a better way to do built-in `+` and `*` functions
  - [x] Write more `LispExpr.parse()` functions.
  - [ ] Think of a better name for `LispExpr.parse_type()`. It's not actually a
    parser.
  - [ ] Refactor parser to directly build a `LispExpr::List`
- [x] Better error handling.
- [ ] More general function definition. Fixed arity functions. Typechecking
  arguments.
  - [ ] Figure out what type should the function be and return.
        Is it a `LispExpr`? Is it a `LispFunc`. In `+` etc. we're evaluating the
        function there but not while composing functions. This causes type
        mismatch.
- [ ] More tests.
  - [ ] More parser tests.
- [ ] PARSER
  - [x] Basic string -> token tree parser.
  - [x] Parse tokens into `LispExpr` tree (ast).
  - [ ] Parser errors
- [x] Write a REPL.
  - [ ] Unwrap results before printing.
  - [ ] Better error messages.
- [ ] Handle special conditions.
- [ ] Add more things to default environment.


history.txt
#V2
(+ 2 (+ 3 4))
(+ (* 3 3) (* 4 4))


main.py
class LispAtom:
    def __init__(self, val):
        self.val = val
    def eval(self, env):
        return self.val

class LispExpr:
    """Either a list of LispExpr or a LispExpr or a LispAtom"""
    def __init__(self, expr):
        self.expr = expr
    def eval(self, env):
        return LispAtom(self.expr[0].eval(env)(
            *map(lambda e: e.eval(env), self.expr[1:])))

class LispEnv:
    def __init__(self, parent=None):
        self.parent = parent
        self.table = {} # dict of (string, LispAtom)
    
    def get(self, name):
        if name in self.table.keys():
            return self.table[name]
        if not self.parent:
            return None
        return self.parent.get(name)
    
    def add(self, name, expr):
        self.table[name] = expr

class LispName:
    """Instances of this are variables"""
    def __init__(self, name):
        self.name = name
    def eval(self, env):
        return env.get(self.name).eval(env)

class LispFunc:
    """This should be a LispName"""
    def __init__(self, args: list, body: LispExpr):
        self.args = args
        self.body = body

    def eval(self, env: LispEnv, params: dict):
        # this is the tricky part. we create a new child environment here
        local_env = LispEnv(parent=env)
        for a, p in params.items():
            local_env.add(a, p)
        return self.body.eval(local_env)

env_global = LispEnv()

env_global.add('+', LispAtom(lambda x, y: x + y))
env_global.add('-', LispAtom(lambda x, y: x - y))
env_global.add('*', LispAtom(lambda x, y: x * y))


expr = LispExpr([LispAtom(lambda x, y: x + y), LispName("x"), LispName("y")])

print(expr.eval(env_global).eval(env_global))


src/lib.rs
use std::collections::HashMap;

type ArgsType = Vec<LispExpr>; // all lisp function take a list of arguments
type ReturnType = Result<LispExpr, LispErr>; // all lisp functions return this type

#[derive(Clone, Debug, PartialEq)]
struct LispFunc {
    func: fn(ArgsType) -> ReturnType,
    arity: usize, // number of arguments
    inf_args: bool,
}

#[derive(Clone, Debug, PartialEq)]
enum LispExpr {
    Symbol(String),
    Integer(i64),
    List(Vec<LispExpr>),
    Func(LispFunc)
}


impl LispExpr {
    fn parse_symbol(&self) -> Result<String, LispErr> {
        match self {
            LispExpr::Symbol(s) => Ok(s.clone()),
            _ => Err(LispErr::TypeError),
        }
    }
    fn parse_int(&self) -> Result<i64, LispErr> {
        match self {
            LispExpr::Integer(i) => Ok(i.clone()),
            _ => Err(LispErr::TypeError),
        }
    }

    fn parse_fn(&self) -> Result<LispFunc, LispErr> {
        match self {
            LispExpr::Func(f) => Ok(f.clone()),
            _ => Err(LispErr::TypeError),
        }
    }
}

#[derive(Debug)]
enum LispErr {
    ArityMismatch,
    NameError,
    TypeError,
}

fn eval(expr: &LispExpr, env: &mut LispEnv) -> ReturnType {
    match expr {
        LispExpr::Symbol(s) => match env.data.get(s) {
            Some(e) => Ok(e.clone()),
            None => Err(LispErr::NameError),
        },
        LispExpr::Integer(_) => Ok(expr.clone()),
        LispExpr::List(list) => {
            let f = eval(&list[0], env)?.parse_fn()?;
            // TODO does slicing here create a copy?
            let args = list[1..].iter()
                .map(|a| eval(a, env))
                .collect::<Result<Vec<_>, _>>()?;

            if !f.inf_args && f.arity != args.len() {
                return Err(LispErr::ArityMismatch)
            }

            // recursively evaluate until the return value is not an atom
            eval(&(f.func)(args)?, env) // TODO who owns the value here?
        },
        LispExpr::Func(_) => Ok(expr.clone()),
    }
}


struct LispEnv {
    data: HashMap<String, LispExpr>,
}

impl LispEnv {
    fn new() -> LispEnv {
        LispEnv { data: HashMap::new() }
    }

    fn default() -> LispEnv {
        let mut env = LispEnv::new();

        use LispExpr::*;

        // addition
        env.insert(
            "+".to_string(),
            Func(LispFunc {
                func: |args: ArgsType| -> ReturnType {
                    // TODO too many copies?
                    // TODO find a better way to do this
                    let ans = args.iter()
                        .map(|a| a.parse_int() )
                        .collect::<Result<Vec<_>, _>>()?
                        .iter()
                        .cloned()
                        .sum();
                    Ok(Integer(ans))
                },
                inf_args: true,
                arity: 0,
            })
        );

        // subtraction
        env.insert(
            "-".to_string(),
            Func(LispFunc {
                func: |args: ArgsType| -> ReturnType {
                    let first = args[0].parse_int()?;
                    let ans: i64 = args[1..].iter()
                        .map(|a| a.parse_int() )
                        .collect::<Result<Vec<_>, _>>()?
                        .iter()
                        .cloned()
                        .sum();
                    Ok(Integer(first - ans))
                },
                inf_args: true,
                arity: 0,
            })
        );

        // multiplication
        env.insert(
            "*".to_string(),
            Func(LispFunc {
                func: |args: ArgsType| -> ReturnType {
                    let ans = args.iter()
                        .map(|a| a.parse_int() )
                        .collect::<Result<Vec<_>, _>>()?
                        .iter()
                        .cloned()
                        .product();
                    Ok(Integer(ans))
                },
                inf_args: true,
                arity: 0,
            })
        );

        env.insert(
            "square".to_string(),
            Func(LispFunc {
                func: |args: ArgsType| -> ReturnType {
                    // TODO remove these clones
                    // TODO should this be wrapped in an Ok?
                    Ok(List(vec![Symbol("*".to_string()),
                                args[0].clone(), args[0].clone()]))
                },
                inf_args: false,
                arity: 1,
            })
        );

        env.insert(
            // always fails
            "bad-func".to_string(),
            Func(LispFunc {
                func: |args: ArgsType| -> ReturnType {
                    Ok(List(vec![Symbol("not-defined".to_string())]))
                },
            inf_args: false,
            arity: 0
            })
        );

        env
    }

    fn insert(&mut self, name: String, expr: LispExpr) {
        self.data.insert(name, expr);
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use LispExpr::*;

    #[test]
    fn eval_arithmetic() {
        let mut env = LispEnv::default();

        let expr = List(vec![Symbol(String::from("+")),
                    List(vec![Symbol(String::from("+")), Integer(3), Integer(5)]),
                    Integer(4)]);
        assert_eq!(eval(&expr, &mut env).unwrap(), Integer(12));

        let expr = List(vec![Symbol(String::from("+")),
                    List(vec![Symbol(String::from("-")), Integer(3), Integer(5)]),
                    Integer(4)]);
        assert_eq!(eval(&expr, &mut env).unwrap(), Integer(2));

        let expr = List(vec![Symbol(String::from("*")),
                    List(vec![Symbol(String::from("+")), Integer(3), Integer(5)]),
                    Integer(4)]);
        assert_eq!(eval(&expr, &mut env).unwrap(), Integer(32));
    }

    #[test]
    fn eval_functions() {
        let mut env = LispEnv::default();

        let expr = List(vec![Symbol("square".to_string()), Integer(5)]);
        assert_eq!(eval(&expr, &mut env).unwrap(), Integer(25));

        let expr = List(vec![Symbol("bad-func".to_string())]);
        assert!(eval(&expr, &mut env).is_err());

        let expr = List(vec![Symbol("square".to_string()),
                             Integer(3),
                             Integer(4)]);

        // TODO write better tests
        assert!(match eval(&expr, &mut env) {
                    Err(LispErr::ArityMismatch) => true,
                    _ => false,
        });
    }
}


src/main.rs
mod lisp;
mod parser;

use lisp::LispEnv;

use rustyline::error::ReadlineError;
use rustyline::Editor;

fn main() {
    let mut env = LispEnv::default();

    // `()` can be used when no completer is required
    let mut rl = Editor::<()>::new();

    loop {
        let readline = rl.readline("> ");
        match readline {
            Ok(line) => {
                println!("Input: {}", line);
                rl.add_history_entry(line.as_str());

                let whitespaced = line.replace("(", " ( ").replace(")", " ) ");
                let mut iter = whitespaced.split_whitespace();
                let parsed = parser::parse(&mut iter);
                let expr = parsed.to_lispexpr().extract_first();

                println!("Parsed Expression: {:?}", expr);
                println!("Result: {:?}", expr.eval(&mut env));
            }
            Err(ReadlineError::Interrupted) => {
                println!("CTRL-C");
                break;
            }
            Err(ReadlineError::Eof) => {
                println!("CTRL-D");
                break;
            }
            Err(err) => {
                println!("Error: {:?}", err);
                break;
            }
        }
    }

    rl.save_history("history.txt").unwrap();
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn parse_test() {
        let inp = "(+ (+ 2 3) (+ 4 5 (+ 6 7))".to_string();
        let whitespaced = inp.replace("(", " ( ").replace(")", " ) ");
        let mut iter = whitespaced.split_whitespace();

        let parsed = parser::parse(&mut iter);
        let mut env = lisp::LispEnv::default();
        let expr = parsed.to_lispexpr().extract_first();

        println!("{:?}", expr);
        println!("{:?}", expr.eval(&mut env).unwrap());
    }
}


src/parser.rs
use crate::lisp::LispExpr;

pub enum AstNode {
    Leaf(LispExpr),
    Node(Vec<AstNode>),
}

impl AstNode {
    #[allow(dead_code)]
    pub fn print(&self) {
        match self {
            AstNode::Leaf(expr) => expr.print(),
            AstNode::Node(v) => {
                print!("(");
                for n in v {
                    n.print();
                }
                print!(")");
            }
        }
    }

    fn push(&mut self, val: AstNode) {
        // TODO handle the case when it might not be a Node
        // TODO prevent that case from happening altogether
        #[allow(clippy::single_match)]
        match self {
            AstNode::Node(v) => v.push(val),
            _ => (),
        }
    }

    // TODO this is redundant.
    // TODO i should directly build the AST as a LispExpr
    pub fn to_lispexpr(&self) -> LispExpr {
        match self {
            AstNode::Leaf(e) => e.clone(),
            AstNode::Node(v) => {
                LispExpr::List(v.iter().map(|e| e.to_lispexpr()).collect::<Vec<_>>())
            }
        }
    }
}

#[allow(dead_code)]
type ParseErr = String;

fn to_lispexpr(token: &str) -> LispExpr {
    match token.parse::<i64>() {
        Ok(num) => LispExpr::Integer(num),
        Err(_) => LispExpr::Symbol(token.to_string()),
    }
}

pub fn parse(iter: &mut dyn Iterator<Item = &str>) -> AstNode {
    let mut ast = AstNode::Node(Vec::new());
    while let Some(token) = iter.next() {
        match token {
            "(" => ast.push(parse(iter)),
            ")" => break,
            _ => ast.push(AstNode::Leaf(to_lispexpr(token))),
        }
    }
    ast
}

#[cfg(test)]
mod tests {
    #[test]
    fn parser_test() {
        assert!(true);
    }
}


src/lisp.rs
use std::collections::HashMap;

type ArgsType = Vec<LispExpr>; // all lisp function take a list of arguments
type ReturnType = Result<LispExpr, LispErr>; // all lisp functions return this type

#[derive(Clone, Debug, PartialEq)]
pub struct LispFunc {
    func: fn(ArgsType) -> ReturnType,
    arity: usize, // number of arguments
    inf_args: bool,
}

#[derive(Clone, Debug, PartialEq)]
pub enum LispExpr {
    Symbol(String),
    Integer(i64),
    List(Vec<LispExpr>),
    Func(LispFunc),
    Bool(bool),
    Null,
}

impl LispExpr {
    #[allow(dead_code)]
    fn extract_symbol(&self) -> Result<String, LispErr> {
        match self {
            LispExpr::Symbol(s) => Ok(s.clone()),
            _ => Err(LispErr::TypeError("expected symbol".to_string())),
        }
    }

    fn extract_int(&self) -> Result<i64, LispErr> {
        match self {
            LispExpr::Integer(n) => Ok(*n),
            _ => Err(LispErr::TypeError("expected integer".to_string())),
        }
    }

    fn extract_fn(&self) -> Result<LispFunc, LispErr> {
        match self {
            LispExpr::Func(f) => Ok(f.clone()),
            _ => Err(LispErr::TypeError("expected func".to_string())),
        }
    }

    fn extract_bool(&self) -> Result<bool, LispErr> {
        match self {
            LispExpr::Bool(b) => Ok(*b),
            _ => Err(LispErr::TypeError("expected bool".to_string())),
        }
    }


    // TODO this shouldn't be required
    // This is needed because the parser creates a List of LispExprs, but it
    // it cannot be evaluated directly because it's not in the proper format
    // (the first element is not a Func)
    pub fn extract_first(&self) -> LispExpr {
        match self {
            LispExpr::List(v) => v[0].clone(),
            _ => self.clone(),
        }
    }

    #[allow(dead_code)]
    pub fn print(&self) {
        match self {
            LispExpr::Symbol(s) => print!("Symbol: {},", s),
            LispExpr::Integer(n) => print!("Integer: {},", n),
            _ => (),
        }
    }

    pub fn eval(&self, env: &mut LispEnv) -> ReturnType {
        match self {
            LispExpr::Symbol(s) => match env.get(s) {
                Some(e) => Ok(e.clone()),
                None => Err(LispErr::NameError),
            },
            LispExpr::Integer(_) => Ok(self.clone()),
            LispExpr::List(list) => {
                if list.len() == 0 {
                    return Ok(LispExpr::Null);
                }

                // handle special forms
                match &list[0].extract_symbol()?.as_str() {
                    &"define" => {
                        if list.len() != 3 {
                            return Err(LispErr::ArityMismatch);
                        }

                        match &list[1] {
                            LispExpr::Symbol(s) => {
                                //let third = list[2].extract_symbol()?;
                                env.insert(s.clone(), list[2].clone());
                                return Ok(LispExpr::Null);
                            }
                            _ => (),
                        }
                    },
                    _ => (),
                }

                let f = &list[0].eval(env)?.extract_fn()?;

                // TODO does slicing here create a copy?
                let args = list[1..]
                    .iter()
                    .map(|a| a.eval(env))
                    .collect::<Result<Vec<_>, _>>()?;

                if !f.inf_args && f.arity != args.len() {
                    return Err(LispErr::ArityMismatch);
                }

                // recursively evaluate until the return value is not an atom
                (f.func)(args)?.eval(env) // TODO who owns the value here?
            }
            LispExpr::Func(_) => Ok(self.clone()),
            LispExpr::Bool(_) => Ok(self.clone()),
            LispExpr::Null => Ok(self.clone()),
        }
    }
}

#[derive(Debug)]
pub enum LispErr {
    ArityMismatch,
    NameError,
    TypeError(String),
}

pub struct LispEnv {
    data: HashMap<String, LispExpr>,
}

impl LispEnv {
    fn new() -> LispEnv {
        LispEnv {
            data: HashMap::new(),
        }
    }

    pub fn default() -> LispEnv {
        let mut env = LispEnv::new();

        use LispExpr::*;

        // addition
        env.insert(
            "+".to_string(),
            Func(LispFunc {
                func: |args: ArgsType| -> ReturnType {
                    // TODO too many copies?
                    // TODO find a better way to do this
                    let ans = args
                        .iter()
                        .map(|a| a.extract_int())
                        .collect::<Result<Vec<_>, _>>()?
                        .iter()
                        .cloned()
                        .sum();
                    Ok(Integer(ans))
                },
                inf_args: true,
                arity: 0,
            }),
        );

        // subtraction
        env.insert(
            "-".to_string(),
            Func(LispFunc {
                func: |args: ArgsType| -> ReturnType {
                    let first = args[0].extract_int()?;
                    let ans: i64 = args[1..]
                        .iter()
                        .map(|a| a.extract_int())
                        .collect::<Result<Vec<_>, _>>()?
                        .iter()
                        .cloned()
                        .sum();
                    Ok(Integer(first - ans))
                },
                inf_args: true,
                arity: 0,
            }),
        );

        // multiplication
        env.insert(
            "*".to_string(),
            Func(LispFunc {
                func: |args: ArgsType| -> ReturnType {
                    let ans = args
                        .iter()
                        .map(|a| a.extract_int())
                        .collect::<Result<Vec<_>, _>>()?
                        .iter()
                        .cloned()
                        .product();
                    Ok(Integer(ans))
                },
                inf_args: true,
                arity: 0,
            }),
        );

        env.insert(
            "square".to_string(),
            Func(LispFunc {
                func: |args: ArgsType| -> ReturnType {
                    // TODO remove these clones
                    // TODO should this be wrapped in an Ok?
                    Ok(List(vec![
                        Symbol("*".to_string()),
                        args[0].clone(),
                        args[0].clone(),
                    ]))
                },
                inf_args: false,
                arity: 1,
            }),
        );

        env.insert(
            // always fails
            "bad-func".to_string(),
            Func(LispFunc {
                func: |_| -> ReturnType { Ok(List(vec![Symbol("not-defined".to_string())])) },
                inf_args: false,
                arity: 0,
            }),
        );

        env
    }

    fn insert(&mut self, name: String, expr: LispExpr) {
        self.data.insert(name, expr);
    }

    fn get(&self, name: &str) -> Option<LispExpr> {
        let res = self.data.get(name)?;
        if let LispExpr::Symbol(new_name) = res {
            self.get(new_name)
        } else {
            Some(res.clone())
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use LispExpr::*;

    #[test]
    fn eval_arithmetic() {
        let mut env = LispEnv::default();

        let expr = List(vec![
            Symbol(String::from("+")),
            List(vec![Symbol(String::from("+")), Integer(3), Integer(5)]),
            Integer(4),
        ]);
        assert_eq!(expr.eval(&mut env).unwrap(), Integer(12));

        let expr = List(vec![
            Symbol(String::from("+")),
            List(vec![Symbol(String::from("-")), Integer(3), Integer(5)]),
            Integer(4),
        ]);
        assert_eq!(expr.eval(&mut env).unwrap(), Integer(2));

        let expr = List(vec![
            Symbol(String::from("*")),
            List(vec![Symbol(String::from("+")), Integer(3), Integer(5)]),
            Integer(4),
        ]);
        assert_eq!(expr.eval(&mut env).unwrap(), Integer(32));
    }

    #[test]
    fn eval_functions() {
        let mut env = LispEnv::default();

        let expr = List(vec![Symbol("square".to_string()), Integer(5)]);
        assert_eq!(expr.eval(&mut env).unwrap(), Integer(25));

        let expr = List(vec![Symbol("bad-func".to_string())]);
        assert!(expr.eval(&mut env).is_err());

        let expr = List(vec![Symbol("square".to_string()), Integer(3), Integer(4)]);

        // TODO write better tests
        assert!(match expr.eval(&mut env) {
            Err(LispErr::ArityMismatch) => true,
            _ => false,
        });
    }

    #[test]
    fn special_forms() {
        let mut env = LispEnv::default();
        let expr = List(vec![
            Symbol("define".to_string()),
            Symbol("add".to_string()),
            Symbol("+".to_string()),
        ]);
        expr.eval(&mut env).unwrap();

    }
}
